### PR TITLE
[PBI-731] Implement a timeout for fw_onAdRequestComplete event

### PR DESCRIPTION
**Pull request by dchauOoyala**
- Adds a timeout to check if there is a response for the ad list request
- Raises error if there is no response

### CODE REVIEW

**Comment by pilievOoyala** [comment]
The timeout needs to be clear on successful callback and when we delete/destroy the plugin.

File: js/freewheel.js [hunk 1]
```diff
@@ -261,9 +261,31 @@ OO.Ads.manager(function(_, $) {
...
+          OO.log("FW: freewheel ad request timeout");
+          amc.raiseAdError("FW: An ad error has occurred. The error string reported was: " + error);
+          slotEndedCallbacks[adRequestType]();
+          delete slotEndedCallbacks[adRequestType];
+        }
+      }, this);
+
       fwContext.submitRequest();
       fwAdDataRequested = true;
+      setAdRequestTimeout(adRequestTimeout, amc.MAX_AD_REQUEST_TIMEOUT);
```
Comment (pilievOoyala) [inline js/freewheel.js]: If the adRequestTimeout already exists, then we are in a bad state.  So throw error.

File: js/freewheel.js [hunk 2]
```diff
@@ -261,9 +263,39 @@ OO.Ads.manager(function(_, $) {
...
+     * if we timeout.
+     * @private
+     * @method Freewheel#_setAdRequestTimeout
+     * @param callback The function to call when we time out
+     * @param duration the time to wait before timing out
+     */
+    var _setAdRequestTimeout = _.bind(function(callback, duration){
+      if (adRequestTimeout) {
+        var error = "Ad Request Timeout already exists - bad state";
+        amc.raiseAdError("FW: An ad error has occurred. The error string reported was: " + error);
```
Comment (pilievOoyala) [inline js/freewheel.js]: You've used this in a couple of places. If you want, you can put this amc.raiseAdError into a function. That way you guarantee the format of the error message will always be the same.  Just a suggestion, not required.

File: js/freewheel.js [hunk 3]
```diff
@@ -283,6 +315,8 @@ OO.Ads.manager(function(_, $) {
       } else {
         OO.log("FW: freewheel metadata request failure");
       }
+      // clear ad request timeout since fw_onAdRequestComplete was called
+      _clearAdRequestTimeout();
```
Comment (pilievOoyala) [inline js/freewheel.js]: Can you move this to the top of the function?  That way if someone ads code and it happens to exit early, we don't forget to do this.

File: js/freewheel.js [hunk 4]
```diff
@@ -261,9 +263,39 @@ OO.Ads.manager(function(_, $) {
...
+      // Set up function to log error if ad response is null after maximum duration allowed for ad request to respond
+      // has exceeded.
+      var adRequestTimeout = _.bind(function(){
+        if (!fwContext._adResponse) {
+          var error = "ad request timeout";
+          OO.log("FW: freewheel ad request timeout");
+          amc.raiseAdError("FW: An ad error has occurred. The error string reported was: " + error);
+          slotEndedCallbacks[adRequestType]();
+          delete slotEndedCallbacks[adRequestType];
+          _clearAdRequestTimeout();
```
Comment (pilievOoyala) [inline js/freewheel.js]: Can you move the clear to be outside the if.  We always want to clear it.

File: js/freewheel.js [hunk 5]
```diff
@@ -261,9 +263,39 @@ OO.Ads.manager(function(_, $) {
         ].join('')
       ];
 
-      // TODO: Implement a timeout here in case fw_onAdRequestComplete doesn't get called - PBI-731
+      // Set up function to log error if ad response is null after maximum duration allowed for ad request to respond
+      // has exceeded.
+      var adRequestTimeout = _.bind(function(){
```
Comment (pilievOoyala) [inline js/freewheel.js]: Is it possible to move this function to be defined out side. In general we are trying to not define functions within functions. (Easier to debug.)

**Comment by pilievOoyala** [comment]
Good work, just a couple of style notes to avoid potential problems in the future.

**Comment by pilievOoyala** [comment]
Looks good to me.  Just make sure to validate that the timeout and all the code won't crash the player and we can still watch the video.

**Comment by pilievOoyala** [comment]
We talked and you validated on desktop and mobile.  Good work!

**Comment by dchauOoyala** [comment]
Hooray!

**Pull request merged**